Add List trait to ModelV2

leovalais opened the pull request:
closes #7031

> [!TIP]
> Better review commit by commit.

https://github.com/OpenRailAssociation/osrd/pull/7350#issuecomment-2086895370

The interface looks like this at usage:

https://github.com/OpenRailAssociation/osrd/blob/2be60fc0cba70351bc3067ef7788ac1e39d2cdd6/editoast/src/modelsv2/mod.rs#L233-L256

leovalais committed:
editoast: fix ModelV2 Patch::apply visibility

leovalais committed:
editoast: add Table associated type to Model trait

leovalais committed:
editoast: use Path instead of string to represent ModelV2 columns

leovalais committed:
editoast: add the codegen of ModelV2's ModelField associated constants

leovalais committed:
editoast: add ModelV2 macro expansion of ModelField settings functions

leovalais commented:
~~@flomonster could you do an intermediate review please? There's `unused` warnings so to silence them I'll replace an `impl modelv1::List` somewhere, but I'd like a review first in order not to redo that part if significant changes are required.~~


I'll do that in another PR.

leovalais committed:
editoast: add temporary allow(unused) in ModelV2 List utils

woshilapin commented on editoast/editoast_derive/src/modelv2/config.rs:
```diff
@@ -152,7 +152,12 @@ impl ModelField {
+            .column
+            .segments
+            .last()
+            .expect("Model: invalid empty column")
```
Do we agree that `.expect()` in a macro will fail at compile time? If yes, you can just resolve this comment.

leovalais replied:
It does, but it won't ;) (I don't believe `syn` successfully parses empty `Path`s)

woshilapin commented on editoast/src/modelsv2/prelude/list.rs:
```diff
@@ -0,0 +1,120 @@
+#[async_trait::async_trait]
+pub trait List: Model {
+    /// Lists and counts the objects that match the provided settings
+    async fn list_and_count(
```
As discussed (but leaving a comment for traceability), let remove that and create 2 different API for `count` and `list`. If later it starts to be a performance problem to have 2 requests, we'll talk again about optimizations.

leovalais replied:
Done, but I left a naive implementation of `list_and_count` that can be customized if necessary.

woshilapin reviewed (changes requested):
Big work 👏 Thank you. Nothing more than what we orally discussed this morning.

leovalais committed:
editoast: add ModelV2 macro expansion for List trait

leovalais committed:
editoast: implement Clone for SelectionSettings

leovalais committed:
editoast: replace list_and_count macro impl in two traits List and Count

- Adds the trait Count
- Splits the complex macro codegen of List::list_and_count in List::list
  and Count::count
- Add trait ListAndCount with a default list_and_count implementation
  that calls [Count::count] and [List::list] sequentially
- Removes the different utils that were useful for the previous
  list_and_count implementation.

woshilapin commented on editoast/src/modelsv2/prelude/list.rs:
```diff
@@ -43,14 +44,17 @@ impl<M: Model> FilterSetting<M> {
 pub struct SelectionSettings<M: Model> {
-    pub(in crate::modelsv2) filters: Vec<FilterSetting<M>>,
-    pub(in crate::modelsv2) sorts: Vec<SortSetting<M>>,
+    pub(in crate::modelsv2) filters: Vec<Arc<dyn Fn() -> FilterSetting<M> + Send + Sync>>,
```
If I understand, we can't have `FilterSetting` be clonabled because of `diesel::BoaxableExpression` which is not.

One way of bypassing that is not to store the object itself, but a factory of the object (a function that can be called to create the object each time it's called).

Another way would be to wrap `FilterSetting` in an `Arc`.

But I do not understand why you do both at the same time (wrapping the factory in an `Arc`). What did I miss?

Also, the function does make the general API less easy to use, this is more straightforward:

```rust
SelectionSettings::new().filter(Document::CONTENT_TYPE.eq(json))
```
than this

```rust
SelectionSettings::new().filter(move || Document::CONTENT_TYPE.eq(json.to_string()))
```

leovalais replied:
The problem is that in the implementation of the traits by the macro, we do:

```rust
for filter_fun in settings.filters {
    let crate::modelsv2::prelude::FilterSetting(filter) = (*filter_fun)();
    query = query.filter(filter);
}
```

in which `query.filter(filter)` requires ownership of `filter` (if we could pass a reference we'd be taking a `&SelectionSettings<Model>` in parameter of the function). So even if `Box<dyn BoxableExpression>` was `Send + Sync`, we'd still need ownership of the pointee of the `Arc<FilterSetting>` to build the expression, thus potentially breaking other copies of the `Arc`. (Or maybe I'm misunderstanding something about `Arc`, which tbh is quite possible 😬)

Even though I agree with you about the API, we'll mostly use `.order_by` which doesn't require any capture so it's literally just an extra `|| ` to pay. I don't believe we'll use `.filter` in many places.
